File: src/net/eduard/api/lib/game/Kit.java
package net.eduard.api.lib.game;

import java.util.ArrayList;
import java.util.List;

import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.PlayerInventory;

import net.eduard.api.lib.modules.Mine;
import net.eduard.api.lib.storage.Storable;

public class Kit   {

    private String name;
    private long cooldown;
    private boolean clearInventory;
    private boolean fillSoup;
    private boolean autoEquip = true;
    private ItemStack icon;
    private int menuPosition;
    private double price;
    private int level;
    private List<ItemStack> items = new ArrayList<>();
    private List<Item> extras = new ArrayList<>();
    private List<Kit> upgrades = new ArrayList<>();

    public Kit(Kit kit, int price, int level) {
        kit.upgrades.add(this);
        setPrice(price);
        setLevel(level);

    }

    public Kit() {

    }

    public void give(Player p) {
        give(p, 1);
    }

    public void give(Player p, int level) {
        PlayerInventory inv = p.getInventory();
        if (clearInventory) {
            Mine.clearInventory(p);
        }
        for (ItemStack item : items) {
            String type = item.getType().name();
            if (autoEquip) {
                if (type.contains("LEGGINGS")) {
                    inv.setLeggings(item);
                } else if (type.contains("CHESTPLATE")) {
                    inv.setChestplate(item);
                } else if (type.contains("BOOTS")) {
                    inv.setBoots(item);
                } else if (type.contains("HELMET")) {
                    inv.setHelmet(item);
                } else {
                    inv.addItem(item);
                }
            } else {
                inv.addItem(item);
            }
        }
        for (int id = 2; id <= level; id++) {
            Kit upgrade = getUpgrade(id);
            for (ItemStack item : upgrade.getItems()) {
                String type = item.getType().name();
                if (autoEquip) {
                    if (type.contains("LEGGINGS")) {
                        inv.setLeggings(item);
                    } else if (type.contains("CHESTPLATE")) {
                        inv.setChestplate(item);
                    } else if (type.contains("BOOTS")) {
                        inv.setBoots(item);
                    } else if (type.contains("HELMET")) {
                        inv.setHelmet(item);
                    } else {
                        inv.addItem(item);
                    }
                } else {
                    inv.addItem(item);
                }
            }
        }
        for (Item item : extras) {
            inv.addItem(item.create());
        }
        if (fillSoup)
            Mine.fill(inv, new ItemStack(Material.MUSHROOM_SOUP));
    }

    public Kit getUpgrade(int level) {
        for (Kit upgrade : upgrades) {
            if (upgrade.getLevel() == level) {
                return upgrade;
            }
        }
        return null;
    }

    public double getPrice() {
        return price;
    }

    public void setPrice(double price) {
        this.price = price;
    }

    public List<ItemStack> getItems() {
        return items;
    }

    public void setItems(List<ItemStack> items) {
        this.items = items;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public ItemStack getIcon() {
        return icon;
    }

    public void setIcon(ItemStack icon) {
        this.icon = icon;
    }

    public boolean isClearInventory() {
        return clearInventory;
    }

    public void setClearInventory(boolean clearInventory) {
        this.clearInventory = clearInventory;
    }

    public boolean isFillSoup() {
        return fillSoup;
    }

    public void setFillSoup(boolean fillSoup) {
        this.fillSoup = fillSoup;
    }

    public long getCooldown() {
        return cooldown;
    }

    public void setCooldown(long cooldown) {
        this.cooldown = cooldown;
    }


    public List<Item> getExtras() {
        return extras;
    }

    public void setExtras(List<Item> extras) {
        this.extras = extras;
    }

    public boolean isAutoEquip() {
        return autoEquip;
    }

    public void setAutoEquip(boolean autoEquip) {
        this.autoEquip = autoEquip;
    }

    public int getLevel() {
        return level;
    }

    public void setLevel(int level) {
        this.level = level;
    }

    public int getMenuPosition() {
        return menuPosition;
    }

    public void setMenuPosition(int menuPosition) {
        this.menuPosition = menuPosition;
    }
}

File: src/main/java/net/eduard/api/lib/game/Kit.kt
package net.eduard.api.lib.game

import net.eduard.api.lib.modules.Mine
import net.eduard.api.server.minigame.MinigameMode
import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.inventory.ItemStack
import java.util.concurrent.TimeUnit

class Kit() {

    constructor(block: Kit.() -> Unit) : this() {
        block.invoke(this)
    }

    class KitUpgrade(
        var level: Int = 1,
        var price: Double = 0.0,
        var cooldown: Long = TimeUnit.HOURS.toMillis(1)
    ) {
        var items = mutableListOf<ItemStack>()

        @Transient
        lateinit var kit: Kit

    }

    fun newUpgrade(level: Int, block: (KitUpgrade.() -> Unit)? = null): KitUpgrade {
        val upgrade = KitUpgrade(level)
        upgrade.kit = this
        upgrades.add(upgrade)
        block?.invoke(upgrade)
        return upgrade
    }

    var name = "kit"
    var permission = "kits.kit"

    @Transient
    var mode: MinigameMode? = null
    var menuPosition = 10
    var cooldown: Long = TimeUnit.HOURS.toMillis(1)
    var price = 1000.0
    var isClearInventory = false
    var isFillSoup = false
    var isAutoEquip = true
    var icon: ItemStack = ItemBuilder(Material.DIAMOND_SWORD).name("§aKit")
    var items = mutableListOf<ItemStack>()
    var upgrades = mutableListOf<KitUpgrade>()

    fun deliverAutoEquip(player : Player, item : ItemStack){
        if (item.type == Material.AIR)return
        val playerInventory = player.inventory
        val type = item.type.name
        if (type.contains("LEGGINGS")) {
            if (playerInventory.leggings == null) {
                playerInventory.leggings = item
            } else {
                deliver(player, item)
            }
        } else if (type.contains("CHESTPLATE")) {
            if (playerInventory.chestplate == null) {
                playerInventory.chestplate = item
            } else {
                deliver(player, item)
            }
        } else if (type.contains("BOOTS")) {
            if (playerInventory.boots == null) {
                playerInventory.boots = item
            } else {
                deliver(player, item)
            }

        } else if (type.contains("HELMET")) {
            if (playerInventory.helmet == null) {
                playerInventory.helmet = item
            } else {
                deliver(player, item)
            }
        } else {
            deliver(player, item)
        }
    }
    fun deliver(player: Player, item: ItemStack) {
        if (item.type == Material.AIR)return
        if (Mine.isFull(player.inventory)) {
            player.world.dropItemNaturally(player.location, item)
        } else {
            player.inventory.addItem(item)
        }
    }
    fun give(player: Player) {
        give(player,1)
    }
    fun give(player: Player, level: Int ) {
        val playerInventory = player.inventory
        if (isClearInventory) {
            Mine.clearInventory(player)
        }
        if (level == 1) {
            for (item in items) {
                if (isAutoEquip) {
                    deliverAutoEquip(player,item)
                } else {
                    deliver(player, item)
                }
            }
        } else {
            val upgrade = getUpgrade(level)
            for (item in upgrade.items) {
                if (isAutoEquip) {
                   deliverAutoEquip(player,item)
                } else {
                    deliver(player, item)
                }
            }
        }
        if (isFillSoup) Mine.fill(playerInventory, ItemStack(Material.MUSHROOM_SOUP))
    }

    fun getUpgrade(level: Int) = upgrades.first {
        it.level == level
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Kit

        if (name != other.name) return false
        if (mode != other.mode) return false

        return true
    }

    override fun hashCode(): Int {
        var result = name.hashCode()
        result = 31 * result + (mode?.hashCode() ?: 0)
        return result
    }

}